compiler/src/source_type.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+// Copyright 2020-2021 postUI Lab. All rights reserved. MIT license.
+
+use serde::{Serialize, Serializer};
+use std::{
+  fmt,
+  path::{Path, PathBuf},
+};
+
+#[repr(i32)]
+#[derive(Clone, Copy, Eq, PartialEq, Debug)]
+pub enum SourceType {
+  JavaScript = 0,
+  JSX = 1,
+  TypeScript = 2,
+  TSX = 3,
+  Json = 4,
+  Wasm = 5,
+  Unknown = 9,
+}
+
+impl fmt::Display for SourceType {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    let value = match self {
+      SourceType::JavaScript => "JavaScript",
+      SourceType::JSX => "JSX",
+      SourceType::TypeScript => "TypeScript",
+      SourceType::TSX => "TSX",
+      SourceType::Json => "Json",
+      SourceType::Wasm => "Wasm",
+      SourceType::Unknown => "Unknown",
+    };
+    write!(f, "{}", value)
+  }
+}
+
+impl<'a> From<&'a Path> for SourceType {
+  fn from(path: &'a Path) -> Self {
+    SourceType::from_path(path)
+  }
+}
+
+impl<'a> From<&'a PathBuf> for SourceType {
+  fn from(path: &'a PathBuf) -> Self {
+    SourceType::from_path(path)
+  }
+}
+
+impl<'a> From<&'a String> for SourceType {
+  fn from(specifier: &'a String) -> Self {
+    SourceType::from_path(&PathBuf::from(specifier))
+  }
+}
+
+impl Default for SourceType {
+  fn default() -> Self {
+    SourceType::Unknown
+  }
+}
+
+impl SourceType {
+  fn from_path(path: &Path) -> Self {
+    match path.extension() {
+      None => SourceType::Unknown,
+      Some(os_str) => match os_str.to_str() {
+        Some("ts") => SourceType::TypeScript,
+        Some("tsx") => SourceType::TSX,
+        Some("js") => SourceType::JavaScript,
+        Some("jsx") => SourceType::JSX,
+        Some("mjs") => SourceType::JavaScript,
+        Some("json") => SourceType::Json,
+        Some("wasm") => SourceType::Wasm,
+        _ => SourceType::Unknown,
+      },
+    }
+  }
+}
+
+impl Serialize for SourceType {
+  fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+  where
+    S: Serializer,
+  {
+    let value = match self {
+      SourceType::JavaScript => 0 as i32,
+      SourceType::JSX => 1 as i32,
+      SourceType::TypeScript => 2 as i32,
+      SourceType::TSX => 3 as i32,
+      SourceType::Json => 4 as i32,
+      SourceType::Wasm => 5 as i32,
+      SourceType::Unknown => 9 as i32,
+    };
```

Comment:
This can be redone with just an `as i32` after the `match` statement. Idk how to use GH suggested change so here is what I am talking about:
```rust
let value = match self {
	SourceType::JavaScript => 0,
	SourceType::JSX => 1,
	SourceType::TypeScript => 2,
	SourceType::TSX => 3,
	SourceType::Json => 4,
	SourceType::Wasm => 5,
	SourceType::Unknown => 9
} as i32;
```

Reply:
ah, looks great

Reply:
@ije If your fine, I can push the spelling corrections and this refactor on this branch.

Reply:
go ahead!